packages/internal/bridge/sdk/src/tokenBridge.ts

```diff
@@ -547,143 +477,640 @@ export class TokenBridge {
+    recipient: string,
+    token: FungibleToken,
+    amount: ethers.BigNumber,
+  ): Promise<DynamicGasEstimatesResponse> {
+    const simulations: Array<any> = [];
+
+    // Encode approval function for non-native tokens.
+    if (token.toUpperCase() !== NATIVE) {
+      // Get erc20 contract
+      const erc20Contract: ethers.Contract = await withBridgeError<ethers.Contract>(
+        async () => new ethers.Contract(token, ERC20, this.config.rootProvider),
+        BridgeErrorType.PROVIDER_ERROR,
+      );
+
+      // Encode function data
+      const txData = await withBridgeError<string>(async () => erc20Contract.interface
+        .encodeFunctionData('approve', [
+          this.config.bridgeContracts.rootERC20BridgeFlowRate,
+          amount,
+        ]), BridgeErrorType.INTERNAL_ERROR);
+
+      simulations.push({
+        network_id: sourceChainId,
+        estimate_gas: true,
+        simulation_type: 'quick',
+        from: sender,
+        to: token,
+        input: txData,
+      });
+    }
+
+    // Get tx data
+    const txData = await this.getTxData(
       sender,
       recipient,
       amount,
       token,
       sourceChainId,
     );
 
-    const txValue = (token.toUpperCase() !== 'NATIVE')
-      ? fees.bridgeFee.toString() : amount.add(fees.bridgeFee).toString();
+    // tx value for simulation mocked as amount + 1 wei for a native bridge and 1 wei for token bridges
+    const txValue = (token.toUpperCase() !== NATIVE) ? '1' : amount.add('1').toString();
 
-    return {
-      feeData: fees,
-      unsignedTx: {
-        data,
-        to: getContractRes.contractAddress,
-        value: txValue,
-        chainId: parseInt(sourceChainId, 10),
-      },
-    };
+    simulations.push({
+      network_id: sourceChainId,
+      estimate_gas: true,
+      simulation_type: 'quick',
+      from: sender,
+      to: this.config.bridgeContracts.rootERC20BridgeFlowRate,
+      input: txData,
+      value: txValue,
+    });
+
+    let axiosResponse:AxiosResponse;
+    const tenderlyAPI = this.getTenderlyEndpoint(sourceChainId);
+    try {
+      axiosResponse = await axios.post(
+        tenderlyAPI,
+        {
+          simulations,
+        },
+        {
+          headers: {
+            'Content-Type': 'application/json',
+          },
+        },
+      );
+    } catch (error: any) {
+      axiosResponse = error.response;
+    }
+
+    if (axiosResponse.data.error) {
+      throw new BridgeError(
+        `Estimating gas failed with the reason: ${axiosResponse.data.error.message}`,
+        BridgeErrorType.TENDERLY_GAS_ESTIMATE_FAILED,
+      );
+    }
+
+    if (axiosResponse.data.simulation_results.length !== simulations.length) {
+      throw new BridgeError(
+        'Estimating gas failed with mismatched responses',
+        BridgeErrorType.TENDERLY_GAS_ESTIMATE_FAILED,
+      );
+    }
+
+    const tenderlyGasEstimatesRes = {} as DynamicGasEstimatesResponse;
+    const simResults = axiosResponse.data.simulation_results;
```

Comment:
This could be moved out into a separate function e.g. `submitTenderlySimulations`.